content/4_Gold/PURS.mdx

```diff
@@ -160,241 +160,300 @@ public class segtree {
+		assert(0 <= ind && ind < len);
+		for (segtree[ind += len] = val; ind > 1; ind >>= 1) {
+			segtree[ind >> 1] = Math.min(segtree[ind], segtree[ind ^ 1]);
 		}
+	}
 
-		for (int i = 0; i < q; i++) {
-			int type = io.nextInt();
-			if (type == 1) {
-				int k = io.nextInt()-1;
-				int u = io.nextInt();
-				seg.set(k, u);
-			} else {
-				int a = io.nextInt()-1;
-				int b = io.nextInt()-1;
-				io.println(seg.min(a, b));
+	/** @return the minimum of all elements in [start, end). */
+	public int rangeMin(int start, int end) {
+		assert(0 <= start && start < len && 0 < end && end <= len);
+		int min = Integer.MAX_VALUE;
+		for (start += len, end += len; start < end; start >>= 1, end >>= 1) {
+			if ((start & 1) != 0) {
+				min = Math.min(min, segtree[start++]);
+			}
+			if ((end & 1) != 0) {
+				min = Math.min(min, segtree[--end]);
 			}
 		}
-		io.close();
+		return min;
 	}
 
-	public static class SegmentTree {
-		private int[] tree;
-		private int n;
-
-		public SegmentTree(int n) {
-			this.n = n;
-			tree = new int[n*2];
-		}
+	public static void main(String[] args) {
+		Kattio io = new Kattio();
+		int arrLen = io.nextInt();
+		int queryNum = io.nextInt();
 
-		public int min(int a, int b) {
-			a += n;
-			b += n;
-			int min = Integer.MAX_VALUE;
-			while (a <= b) {
-				if (a%2 == 1) min = Math.min(min, tree[a++]);
-				if (b%2 == 0) min = Math.min(min, tree[b--]);
-				a /= 2;
-				b /= 2;
-			}
-			return min;
+		MinSegmentTree segtree = new MinSegmentTree(arrLen);
+		for (int i = 0; i < arrLen; i++) {
+			segtree.set(i, io.nextInt());
 		}
 
-		public void set(int index, int val) {
-			index += n;
-			tree[index] = val;
-			for (index /= 2; index >= 1; index /= 2) {
-				tree[index] = Math.min(tree[2*index], tree[2*index+1]);
+		for (int i = 0; i < queryNum; i++) {
+			int type = io.nextInt();
+			if (type == 1) {
+				int ind = io.nextInt() - 1;
+				int val = io.nextInt();
+				segtree.set(ind, val);
+			} else {
+				int start = io.nextInt() - 1;
+				int end = io.nextInt() - 1;
+				io.println(segtree.rangeMin(start, end + 1));
 			}
 		}
+		io.close();
 	}
 
-	CodeSnip{Kattio}
+	//CodeSnip{Kattio}
 }
 ```
 
 </JavaSection>
-
 </LanguageSection>
 
 ### Solution - Dynamic Range Sum Queries
 
 <LanguageSection>
-
 <CPPSection>
 
-Compared to the previous problem, all we need to change are `T`, `ID`, and
-`comb`.
+Compared to the previous problem, all we need to change are `DEFAULT` and `comb`.
 
 ```cpp
-#include <bits/stdc++.h>
-using namespace std;
+#include <iostream>
+#include <cassert>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+template<class T>
+class SumSegmentTree {
+	private:
+		T comb(T a, T b) { return a + b; }
+		const T DEFAULT = 0;
+
+		vector<T> segtree;
+		int len;
+	public:
+		SumSegmentTree(int len) : len(len), segtree(len * 2, DEFAULT) { }
+
+		void set(int ind, T val) {
+			assert(0 <= ind && ind < len);
+			for (segtree[ind += len] = val; ind > 1; ind >>= 1) {
```

Comment:
I personal find `ind /= 2` more clear, but that's up to u
and maybe pull one of the assignment out of the for parentheses, the order of operations might be confusing

Reply:
tru tru